starlark/eval.go

```diff
@@ -478,26 +478,18 @@ func setIndex(fr *Frame, x, y, z Value) error {
 
 // Unary applies a unary operator (+, -, ~, not) to its operand.
 func Unary(op syntax.Token, x Value) (Value, error) {
-	switch op {
-	case syntax.MINUS:
-		switch x := x.(type) {
-		case Int:
-			return zero.Sub(x), nil
-		case Float:
-			return -x, nil
-		}
-	case syntax.PLUS:
-		switch x.(type) {
-		case Int, Float:
-			return x, nil
-		}
-	case syntax.TILDE:
-		if xint, ok := x.(Int); ok {
-			return xint.Not(), nil
-		}
-	case syntax.NOT:
+	if op == syntax.NOT {
 		return !x.Truth(), nil
 	}
+
+	// Int, Float, and user-defined types
+	if x, ok := x.(HasUnary); ok {
+		y, err := x.Unary(op)
+		if y != nil || err != nil {
```

Comment:
Should this be `y == nil`?

Reply:
Oh, I see later that `nil, nil` has a special meaning. This deserves a comment here, because it definitely reads like a typo.

Reply:
Done.